### PR TITLE
Correcting plasticity CL (and readjusting the dependent CL's)

### PR DESCRIPTION
**Pull request by SergioJimenezReyes**
**Description**
I have observed some aspects to be fixed from the small-strain plasticity CL. This affects all the CL's inherited from plasticity CL, i.e. (small and finite strain) kinematic/isotropic plasticity and the corresponding tests.

**Changelog**
Changes are mainly located in the CL's integrators. Main ones:
- Computation of hardening parameter which is a function of G_flux and not F_flux
- Internal variables on the computation of the equivalent stress threshold
- Readapting the cpp tests to the changes

### CODE REVIEW

**Review by AlejandroCornejo — COMMENTED** [review]
Thank you @SergioJimenezReyes for your revision. The python tests have been updated?

**Comment by loumalouomega** [comment]
Thanks @SergioJimenezReyes I trust the revision to @AlejandroCornejo

**Comment by SergioJimenezReyes** [comment]
> Thank you @SergioJimenezReyes for your revision. The python tests have been updated?

Nope, I'm working on it now.

**Comment by SergioJimenezReyes** [comment]
> Thank you @SergioJimenezReyes for your revision. The python tests have been updated?

Now python tests are updated! (and all checks have passed)

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 1]
```diff
@@ -269,16 +269,16 @@ class GenericConstitutiveLawIntegratorKinematicPlasticity
         CalculateIndicatorsFactors(rPredictiveStressVector, tensile_indicator_factor,compression_indicator_factor);
         CalculatePlasticDissipation(rPredictiveStressVector, tensile_indicator_factor,compression_indicator_factor, rPlasticStrainIncrement,rPlasticDissipation, h_capa, rValues, CharacteristicLength);
         CalculateEquivalentPlasticStrain(rPredictiveStressVector, rUniaxialStress, rPlasticStrain, tensile_indicator_factor, rValues, equivalent_plastic_strain);
-        CalculateEquivalentStressThreshold(rPlasticDissipation, tensile_indicator_factor,compression_indicator_factor, rThreshold, slope, rValues, equivalent_plastic_strain);
-        CalculateHardeningParameter(rYieldSurfaceDerivative, slope, h_capa, hardening_parameter);
+        CalculateEquivalentStressThreshold(rPlasticDissipation, tensile_indicator_factor,compression_indicator_factor, rThreshold, slope, rValues, equivalent_plastic_strain, CharacteristicLength);
+        CalculateHardeningParameter(rDerivativePlasticPotential, slope, h_capa, hardening_parameter);
```
Comment (AlejandroCornejo) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:273]: nice

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 2]
```diff
@@ -505,7 +506,8 @@ class GenericConstitutiveLawIntegratorKinematicPlasticity
         double& rEquivalentStressThreshold,
         double& rSlope,
         ConstitutiveLaw::Parameters& rValues,
-        const double EquivalentPlasticStrain
+        const double EquivalentPlasticStrain,
+        const double CharacteristicLength
```
Comment (AlejandroCornejo) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:510]: why do you need the Lchar now

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 3]
```diff
@@ -610,15 +612,15 @@ class GenericConstitutiveLawIntegratorKinematicPlasticity
         GetInitialUniaxialThreshold(rValues, initial_threshold);
 
         rEquivalentStressThreshold = initial_threshold * (1.0 - PlasticDissipation);
-        rSlope = -0.5 * initial_threshold;
+        rSlope = - initial_threshold;
```
Comment (AlejandroCornejo) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:615]: nice

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 4]
```diff
@@ -694,29 +697,37 @@ class GenericConstitutiveLawIntegratorKinematicPlasticity
...
-        const double EquivalentPlasticStrain
+        const double EquivalentPlasticStrain,
+        const double CharacteristicLength
         )
     {
         const Properties& r_material_properties = rValues.GetMaterialProperties();
         const Vector& curve_fitting_parameters = r_material_properties[CURVE_FITTING_PARAMETERS];
+
+        const bool has_tangency_linear_region = r_material_properties.Has(TANGENCY_REGION2);
+        const bool& tangency_linear_region = has_tangency_linear_region ? r_material_properties[TANGENCY_REGION2] : false;
```
Comment (AlejandroCornejo) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h]: not necessary to put the & (only with Vector and Matrix :))

**Review by philbucher — DISMISSED** [review]
👍

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h (hunk 2 above)
Comment (SergioJimenezReyes) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:510]: It is needed in the _CurveFittingHardening_ curve type for computing the volumetric fracture energy g_f = G_f / Lchar. I am not sure if it is better to recalculated it or to bring it through the methods (it is computed outside the integrator)

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:510]: it is okkey then

**Review by AlejandroCornejo — APPROVED** [review]

**Pull request merged**